### PR TITLE
Guard URI percent decoding bounds

### PR DESCRIPTION
**Pull request by jkwak-work**
## Summary of the problem from the end user perspective

Malformed file URIs from an IDE client could cause the language server to read beyond the URI string while decoding a trailing percent escape, or decode invalid percent escapes into garbage path characters.

## Root cause

`URI::getPath()` decoded every `%` as a three-character escape without first checking that two valid hex digits remained before the path or query boundary.

## Solution in this PR

Only decode a percent escape when both following characters are still within the URI path and are valid hex digits. Too-short or invalid escapes are preserved literally, and local-file URI encoding now emits two-digit hex escapes so encoder/decoder round trips remain consistent.

### Notes to the reviewers; where to focus on

The implementation change is the added guard in `URI::getPath()` plus two-digit padding in `URI::fromLocalFilePath()`. The unit test covers normal `%20` decoding, trailing `%`, trailing `%2`, invalid `%2g`, a `%2` query-boundary case, and a single-digit-hex local path round trip.

### CODE REVIEW

**Comment by jkwak-work** [comment]
@coderabbitai review

File: source/core/slang-io.cpp [hunk 1]
```diff
@@ -1244,7 +1244,7 @@ String URI::getPath() const
     for (Index i = startIndex; i < endIndex;)
     {
         auto ch = uri[i];
-        if (ch == '%')
+        if (ch == '%' && i + 2 < endIndex)
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-io.cpp]: [Agent] Addressed in `1464aae65`: `URI::getPath()` now requires both characters after `%` to be valid hex digits before decoding. Invalid escapes are preserved literally. Validation: `build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal`.

File: tools/slang-unit-test/unit-test-io.cpp [hunk 2]
```diff
@@ -165,3 +165,10 @@ SLANG_UNIT_TEST(ioLargeFileExists)
     SLANG_IGNORE_TEST
 #endif
 }
+
+SLANG_UNIT_TEST(uriGetPathPercentDecode)
+{
+    SLANG_CHECK(URI::fromString(toSlice("file://path%20name")).getPath() == "path name");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%")).getPath() == "path%");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%2")).getPath() == "path%2");
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-io.cpp:173]: [Agent] Addressed in `1464aae65`: added a `%2g` regression case to `uriGetPathPercentDecode` to verify invalid hex escapes are preserved literally. Validation: `build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal`.

File: source/core/slang-io.cpp [hunk 3]
```diff
@@ -1244,7 +1244,8 @@ String URI::getPath() const
     for (Index i = startIndex; i < endIndex;)
     {
         auto ch = uri[i];
-        if (ch == '%')
+        if (ch == '%' && i + 2 < endIndex && CharUtil::isHexDigit(uri[i + 1]) &&
+            CharUtil::isHexDigit(uri[i + 2]))
```
Comment (jkwak-work) [inline source/core/slang-io.cpp]: Let's use temporary local variable to avoid calling the same function twice unnecessarily.

**Comment by jkwak-work** [comment]
@coderabbitai review

File: source/core/slang-io.cpp (hunk 3 above)
Comment (jkwak-work) [reply source/core/slang-io.cpp]: [Agent] Addressed in `0d7002b23`: `URI::getPath()` now computes the two hex digit values once, checks them, and reuses them for the decoded character. Validation: `build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal`.

File: tools/slang-unit-test/unit-test-io.cpp [hunk 4]
```diff
@@ -165,3 +165,11 @@ SLANG_UNIT_TEST(ioLargeFileExists)
...
+    SLANG_CHECK(URI::fromString(toSlice("file://path%2g")).getPath() == "path%2g");
+}
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-io.cpp:177]: [Agent] Addressed in `0d7002b23`: added `file://path%2?x=1` coverage so `%2` remains literal at the query boundary. Validation: `build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal`.

**Comment by jkwak-work** [comment]
[Agent] Addressed CodeRabbit outside-diff feedback in `0d7002b23`: `URI::fromLocalFilePath()` now emits two-digit hex escapes, and the URI unit test covers a single-digit-hex local path round trip. Validation: `build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal`.

**Comment by jkwak-work** [comment]
Closing in favor of upstream PR: https://github.com/shader-slang/slang/pull/11515
